Source/CarthageKit/Project.swift

```diff
@@ -187,11 +187,19 @@ public final class Project {
 		}
 
 		return cartfile.zipWith(privateCartfile)
-			.map { (var cartfile, privateCartfile) -> Cartfile in
-				cartfile.appendCartfile(privateCartfile)
+			.map { (var cartfile, privateCartfile) -> ColdSignal<Cartfile> in
+				// TODO: use Set when Carthage supports Swift 1.2
+				let duplicateDeps = filter(privateCartfile.dependencies) { privateDep in
+					contains(cartfile.dependencies, {$0.project == privateDep.project})
+				}
+
+				if duplicateDeps.count == 0 {
+					return .single(cartfile)
+				}
 
-				return cartfile
+				return .error(CarthageError.DuplicateDependenciesInPrivateCartfile(duplicateDeps).error)
 			}
+			.first().value()!
```

Comment:
If the above is [changed to `tryMap`](https://github.com/Carthage/Carthage/pull/364/files#r25580922), this can go away.

Even if it remained `map` though, it'd be better to use something like `merge` here. `first` is a blocking call that should be avoided as much as possible.